src/ser.rs

```diff
@@ -0,0 +1,86 @@
+use crate::*;
+use serde::{
+    de::{Deserializer, Error, MapAccess, Visitor},
+    ser::{SerializeMap, Serializer},
+    Deserialize, Serialize,
+};
+use std::{fmt, marker::PhantomData};
+
+///////////////////////////////////////////// ~ser~~~ /////////////////////////////////////////////
+
+#[derive(Serialize)]
+#[serde(untagged)]
+#[serde(bound(serialize = "T::Kind: Serialize"))]
+enum NodeContentSer<'a, T: Types> {
+    Leaf(&'a str),
+    Branch(&'a [GreenNode<T>]),
+}
+
+impl<T: Types> Serialize for GreenNode<T>
+where
+    T::Kind: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(1))?;
+        if let Some(text) = self.leaf_text() {
+            map.serialize_entry(&self.kind(), &NodeContentSer::<T>::Leaf(text.as_str()))?;
+        } else {
+            map.serialize_entry(&self.kind(), &NodeContentSer::Branch(self.children()))?;
+        }
+        map.end()
+    }
+}
+
+///////////////////////////////////////////// ~~~~de~ /////////////////////////////////////////////
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+#[serde(bound(deserialize = "T::Kind: Deserialize<'de>"))]
+enum NodeContentDe<'a, T: Types> {
+    Leaf(&'a str),
+    Branch(Vec<GreenNode<T>>),
+}
+
+struct NodeVisitor<T: Types>(PhantomData<GreenNode<T>>);
+
+impl<'de, T: Types> Visitor<'de> for NodeVisitor<T>
+where
+    T::Kind: Deserialize<'de>,
+{
+    type Value = GreenNode<T>;
+
+    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "a singleton map from the type kind to node content")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<GreenNode<T>, A::Error>
+    where
+        A: MapAccess<'de>,
+    {
+        let kind: T::Kind = map
+            .next_key()?
+            .ok_or_else(|| A::Error::custom("unexpected empty map"))?;
+        let value: NodeContentDe<T> = map.next_value()?;
```

Comment:
Untagged enums generally have pretty poor error reporting on deserialization. If there was any type error somewhere inside the vec of GreenNode, all we can know is that the input "Isn't" a vec of GreenNode so you get a message that none of the NodeContentDe variants were matched, rather than pinpointing the problem. I don't know how much you care about this but it would be possible to do better by handwriting the Deserialize impl for NodeContentDe to have a visit_str method and visit_seq method.